src/FffGcodeWriter.cpp

```diff
@@ -1455,168 +1456,275 @@ bool FffGcodeWriter::processSkinAndPerimeterGaps(const SliceDataStorage& storage
     const int wall_0_extruder_nr = mesh.getSettingAsExtruderNr("wall_0_extruder_nr");
-    const int64_t z = layer_nr * getSettingInMicrons("layer_height");
 
     const bool fill_perimeter_gaps =
         mesh.getSettingAsFillPerimeterGapMode("fill_perimeter_gaps") != FillPerimeterGapMode::NOWHERE
         && !getSettingBoolean("magic_spiralize")
         && extruder_nr == wall_0_extruder_nr;
-    const bool fill_concentric_perimeter_gaps =
+
+
+    bool added_something = false;
+
+
+    // add roofing
+    const bool fill_roofing_concentric_perimeter_gaps =
         mesh.getSettingAsFillPerimeterGapMode("fill_perimeter_gaps") != FillPerimeterGapMode::NOWHERE
         && !getSettingBoolean("magic_spiralize")
-        && extruder_nr == top_bottom_extruder_nr;
+        && extruder_nr == roofing_extruder_nr;
+    EFillMethod roofing_pattern = mesh.getSettingAsFillMethod("roofing_pattern");
+    const bool generate_roofing_perimeter_gaps = fill_roofing_concentric_perimeter_gaps && roofing_pattern == EFillMethod::CONCENTRIC; // whether we need to generate perimeter gaps for concentric infill of the roofing area
 
-    bool added_something = false;
+    Polygons roofing_concentric_perimeter_gaps; // the perimeter gaps of the insets of concentric skin pattern of this skin part
+    processSkinPartRoofingInfillGeneratePerimeterGaps(storage, gcode_layer, mesh, extruder_nr, mesh_config, skin_part, generate_roofing_perimeter_gaps, roofing_concentric_perimeter_gaps, added_something);
 
-    Polygons skin_polygons;
-    Polygons skin_lines;
 
-    EFillMethod pattern = (layer_nr == 0)?
+    // add normal skinfill
+    const bool fill_top_bottom_concentric_perimeter_gaps =
+        mesh.getSettingAsFillPerimeterGapMode("fill_perimeter_gaps") != FillPerimeterGapMode::NOWHERE
+        && !getSettingBoolean("magic_spiralize")
+        && extruder_nr == top_bottom_extruder_nr;
+    EFillMethod top_bottom_pattern = (gcode_layer.getLayerNr() == 0)?
         mesh.getSettingAsFillMethod("top_bottom_pattern_0") :
         mesh.getSettingAsFillMethod("top_bottom_pattern");
-    int bridge = -1;
-    if (layer_nr > 0)
-    {
-        bridge = bridgeAngle(skin_part.outline, &mesh.layers[layer_nr-1]);
-    }
-    if (bridge > -1)
+    const bool generate_top_bottom_perimeter_gaps = fill_top_bottom_concentric_perimeter_gaps && top_bottom_pattern == EFillMethod::CONCENTRIC; // whether we need to generate perimeter gaps for concentric infill of the skinfill area
+
+    Polygons top_bottom_concentric_perimeter_gaps; // the perimeter gaps of the insets of concentric skin pattern of this skin part
+    processSkinPartInfillGeneratePerimeterGaps(storage, gcode_layer, mesh, extruder_nr, mesh_config, skin_part, generate_top_bottom_perimeter_gaps, top_bottom_concentric_perimeter_gaps, added_something);
+
+
+    // handle perimeter_gaps of concentric skin
+    if (fill_perimeter_gaps || generate_top_bottom_perimeter_gaps)
```

Comment:
I indeed don't think you can have both, with the other constraints of which features will be printed first (e.g. inner before outer wall) and with the hard requirement that there may be at most `extruder_count - 1` extruder switches per layer.